File: RayTracerModelVB/Plane.vb
Public Class Plane
    Inherits Thing

    Public Property Norm As Vector3

    Public Property Offset As Double

    Public Sub New(ByVal norm As Vector3, ByVal offset As Double, ByVal surface As SurfaceTexture)
        MyBase.New(surface)
        Me.Norm = norm
        Me.Offset = offset
    End Sub

    Public Overrides Function CalculateIntersection(ByVal withRay As Ray) As Intersection
        Dim denom As Double = Norm.DotProduct(withRay.Dir)
        If denom > 0 Then Return Nothing
        Return New Intersection(Me, withRay, (Norm.DotProduct(withRay.Start) + Offset) / -denom)
    End Function

    Public Overrides Function CalculateNormal(ByVal surfacePosition As Vector3) As Vector3
        Return Norm
    End Function
End Class

File: RayTracerModel/Plane.cs
namespace RayTracer
{
    public class Plane : Thing
    {
        public Vector3 Norm { get; private set; }
        public double Offset { get; private set; }

        public Plane(Vector3 norm, double offset, SurfaceTexture surface) : base(surface)
        {
            Norm = norm;
            Offset = offset;
        }

        public override Intersection CalculateIntersection(Ray withRay)
        {
            double denom = Norm.DotProduct(withRay.Dir);
            if (denom > 0) return null;
            return new Intersection(this, withRay, (Norm.DotProduct(withRay.Start) + Offset) / (-denom));
        }

        public override Vector3 CalculateNormal(Vector3 surfacePosition)
        {
            return Norm;
        }
    }
}
